core/node/libp2p/routing.go

```diff
@@ -244,7 +246,8 @@ func PubsubRouter(mctx helpers.MetricsCtx, lc fx.Lifecycle, in p2pPSRoutingIn) (
 					Namespaces: []string{"ipns"},
 				},
 			},
-			Priority: 100,
+			Priority:                       100,
+			DoNotWaitForStreamingResponses: true,
```

Comment:
Can't start a thread on an unchanged area, so going here. Doesn't this setup mean we're always waiting for the DHT even if IPNS over PubSub comes back with a valid response because the DHT is not also tagged as `DoNotWaitForStreamingResponses`?

However, we can't just use `DoNotWaitForStreamingResponses` for the DHT since for provider records we want them along with the IPNI records not just one set.

Reply:
> Can't start a thread on an unchanged area, so going here. Doesn't this setup mean we're always waiting for the DHT even if IPNS over PubSub comes back with a valid response because the DHT is not also tagged as `DoNotWaitForStreamingResponses`?

Yes it does, currently that means pubsub only accelerate recursive queries (because the recursive resolver have a rollback implementation).
To get back the old behaviour all of the routers would need to be marked `DoNotWaitForStreamingResponses`, but that sounds wrong to me, we don't want the Lan DHT to cancel (\*if we use a dual DHT instead of composing them then it's less bad because we only have the issue that the HTTP router can cancel the DHT and vice versa).

An alternative would be to have a setup like this (pseudo code), so all routers but the pubsub one can cancel pubsub or pubsub can cancel all of them:
```json
[
  {
    "Router": {
      "Type": "ComposableRouter",
      "Routers": [
        {
          "Router": {
            "Type": "DHTWan"
          }
        },
        {
          "Router": {
            "Type": "DHTLan"
          }
        },
        {
          "Router": {
            "Type": "HTTPRouter"
          }
        }
      ]
    },
    "DoNotWaitForStreamingResponses": true
  },
  {
    "Router": {
      "Type": "PubSub"
    },
    "DoNotWaitForStreamingResponses": true
  }
]
```

Reply:
If we had https://github.com/ipfs/boxo/issues/397 (which is interesting either way), we wouldn't have such problem.